vit-servicing-station-lib/src/server/snapshot_watcher.rs

```diff
@@ -0,0 +1,291 @@
+
+    #[error("snapshot path is not a file with a proper parent directory")]
+    InvalidPath,
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+}
+
+#[must_use]
+pub struct WatcherGuard(notify::RecommendedWatcher);
+
+type Tag = String;
+
+fn extract_tag_from_filename(path: impl AsRef<OsStr>) -> Option<Tag> {
+    let path = path.as_ref().to_string_lossy();
+
+    path.find(&PAT)
+        // check that the pattern is at the end, that means, that there is nothing trailing, just
+        // in case
+        .filter(|start| start + PAT.as_bytes().len() == path.as_bytes().len())
+        .map(|start| path[..start].to_string())
+}
+
+async fn load_from_paths(
+    mut debouncer: Option<&mut HashMap<Tag, Instant>>,
+    paths: impl Iterator<Item = PathBuf>,
+    pool: &DbConnectionPool,
+) {
+    for path in paths {
+        let tag = match path.file_name().and_then(extract_tag_from_filename) {
+            Some(tag) => tag,
+            None => {
+                trace!("skipping {:?}", path);
+                continue;
+            }
+        };
+
+        if let Some(ref mut debouncer) = debouncer {
+            // A simple debounce to avoid useless reloads, since a single write can trigger many
+            // events. We could do something more complex by coalescing events in timeframe
+            // 'buckets', but since we do the same thing regardless of the event I don't see much
+            // point.
+            let now = Instant::now();
+
+            let last_update = debouncer
+                .get(&tag)
+                .copied()
+                // set the last update in the past if the entry doesn't exist, so the we don't
+                // reach the `continue` the first time.
+                .unwrap_or_else(|| now - DEBOUNCE_TIME);
+
+            if now.duration_since(last_update) < DEBOUNCE_TIME {
+                continue;
+            }
+
+            debouncer.insert(tag.clone(), now);
+        }
+
+        match load_snapshot_table_from_file(path, tag.clone(), pool.clone()).await {
+            Ok(inserted) => {
+                // Maybe this logic can be simplified, I'm not sure. This is mostly to not have a
+                // "memory leak", but it is a minor concern since there shouldn't ever be that many
+                // values for that to cause a problem. As an alternative, we could periodically
+                // remove entries that are too old.
+                if inserted == 0 {
+                    if let Some(ref mut debouncer) = debouncer {
+                        let _ = debouncer.remove(&tag);
+                    }
+                }
+            }
+            Err(error) => {
+                error!(
+                    context = "failed to fill snapshot table from file",
+                    %error
+                );
+            }
+        }
+    }
+}
+
+#[tracing::instrument(skip(context))]
+pub async fn async_watch(path: PathBuf, context: SharedContext) -> Result<WatcherGuard, Error> {
```

Comment:
Wondering if this is the best way to handle updates. Seems to carry some additional limitations (need access to the instance filesystem, ) and complexity (e.g. watcher service, selecting interesting notifications, tag parsing, debouncer, ...) compared to a `PUT`/`POST` request  but still requires outside intervention to feed new data in.
What was the thought process behind this? I may be missing some considerations / aspects

Reply:
Tagging @const-iohk here since reading from the fs was mostly his idea.

I think having a new endpoint where we POST the file content would be fine and simpler indeed. Or if you mean just posting something like "reaload this file and use this tag".

One question I have with that is whether we can restrict that endpoint to the public but use it ourselves (at the architecture, I know we could have an api key or something like that). 

Another consideration would be that if we need to restart the service then we need to remember re-posting too, since I think the db is restored in its initial state. But besides this, I don't have any problem with that.

Reply:
>One question I have with that is whether we can restrict that endpoint to the public but use it ourselves (at the architecture, I know we could have an api key or something like that).

In other places we have a proxy for public interaction that only allows some endpoints, maybe we can have something similar here

Reply:
Another alternative, by the way, would be to provide a CLI tool/command that just writes to the DB directly. It would lock the db for a bit, but it should be pretty much insignificant, especially since for this it would be called infrequently. I'd imagine we can run processes on the instance.